### PR TITLE
feat: improve cli output when building project [fixes DXJ-406]

### PR DESCRIPTION
**Pull request by shamsartem**

### CODE REVIEW

File: src/lib/execPromise.ts [hunk 1]
```diff
@@ -37,33 +41,40 @@ export type ExecPromiseArg = {
...
 
-  const commandToDisplay = IS_DEVELOPMENT
-    ? fullCommand
-    : fullCommand.replace(
-        /([\S\s]*--sk ).*([\S\s]*)/,
-        "$1<SECRET_KEY_IS_HIDDEN>$2"
-      );
+  const getCommandFailedMessage = (code: number | null = null) => {
+    return `Command: ${color.yellow(
+      IS_DEVELOPMENT
```
Comment (folex) [inline src/lib/execPromise.ts:60]: How `IS_DEVELOPMENT` is set?

File: src/lib/execPromise.ts [hunk 2]
```diff
@@ -37,33 +41,40 @@ export type ExecPromiseArg = {
...
+        ? fullCommand
+        : fullCommand.replace(
+            /([\S\s]*--sk ).*([\S\s]*)/,
+            "$1<SECRET_KEY_IS_HIDDEN>$2"
+          )
+    )} ${code === null ? "failed" : `exited with code ${code}`}`;
```
Comment (folex) [inline src/lib/execPromise.ts:66]: how could `code` be null?

File: src/lib/execPromise.ts [hunk 3]
```diff
@@ -73,13 +84,11 @@ export const execPromise = ({
...
 
-        rej(
-          new Error(
-            `Execution timed out: command didn't yield any result in ${color.yellow(
-              `${timeout}ms`
-            )}\n${debugInfo}`
-          )
-        );
+        res({
+          errorMessage: `${getCommandFailedMessage()}. Reason: Execution timed out: command didn't yield any result in ${color.yellow(
```
Comment (folex) [inline src/lib/execPromise.ts:88]: > command didn't yield
what command? or will it be inside `getCommandFailedMessage`?

File: src/lib/execPromise.ts [hunk 4]
```diff
@@ -126,32 +140,39 @@ export const execPromise = ({
...
     });
   });
+
+  if ("expectedErrorMessage" in result) {
+    throw new CLIError(result.expectedErrorMessage);
+  }
+
+  if ("errorMessage" in result) {
+    throw new Error(result.errorMessage);
+  }
```
Comment (folex) [inline src/lib/execPromise.ts:175]: What is `expectedErrorMessage` and why is it different from `errorMessage`?

File: src/lib/execPromise.ts [hunk 5]
```diff
@@ -126,32 +140,39 @@ export const execPromise = ({
...
+        res({
+          errorMessage: `${getCommandFailedMessage(code)}${getErrorMessage(
+            printOutput,
+            stderr
+          )}`,
+        });
       }
 
-      res(stdout);
+      res({ success: stdout });
```
Comment (folex) [inline src/lib/execPromise.ts:165]: maybe name it `stdout`?

later you have `return result.success`, and it's pretty hard to guess that `success` means `stdout`.

File: src/lib/execPromise.ts [hunk 6]
```diff
@@ -126,32 +140,39 @@ export const execPromise = ({
...
+
+  if ("expectedErrorMessage" in result) {
+    throw new CLIError(result.expectedErrorMessage);
+  }
+
+  if ("errorMessage" in result) {
+    throw new Error(result.errorMessage);
+  }
+
+  return result.success;
```
Comment (folex) [inline src/lib/execPromise.ts:177]:
```suggestion
  return result.stdout;
```

File: src/lib/marineCli.ts [hunk 7]
```diff
@@ -62,16 +63,28 @@ export const initMarineCli = async (
...
-    });
+    try {
+      return await execPromise({
+        command: marineCLIPath,
+        args,
+        flags,
+        spinnerMessage:
+          message === undefined
+            ? undefined
+            : getMessageWithKeyValuePairs(message, keyValuePairs),
```
Comment (folex) [inline src/lib/marineCli.ts:74]: Let's move that to a separate variable plz? in-place takes 4 lines, making it hard to grasp

File: src/lib/marineCli.ts [hunk 8]
```diff
@@ -62,16 +63,28 @@ export const initMarineCli = async (
...
+        options: { cwd },
+        printOutput,
+      });
+    } catch (e) {
+      if (
+        isObject(e) &&
+        hasKey("message", e) &&
+        typeof e["message"] === "string"
+      ) {
+        return commandObj.error(e["message"]);
```
Comment (folex) [inline src/lib/marineCli.ts:84]: `errorMessage`, `expectedErrorMessage`, `message` – so many names for a conceptually same field 🤔

File: src/lib/rust.ts [hunk 9]
```diff
@@ -285,9 +285,17 @@ const tryDownloadingBinary = async ({
 
   const url = `https://github.com/fluencelabs/marine/releases/download/${name}-v${version}/${name}-${platformToUse}-x86_64`;
 
+  startSpinner(
+    `Trying to download pre-built ${name}@${version} binary to ${replaceHomeDir(
```
Comment (folex) [inline src/lib/rust.ts:289]:
```suggestion
    `Downloading pre-built ${name}@${version} binary to ${replaceHomeDir(
```

Comment (folex) [reply src/lib/rust.ts:289]: I think "prebuilt" is kinda confusing here: user doesn't need to think about how our CI works

File: src/lib/rust.ts [hunk 10]
```diff
@@ -285,9 +285,17 @@ const tryDownloadingBinary = async ({
...
+      dependencyDirPath
+    )}`
+  );
+
   try {
     await downloadFile(binaryPath, url);
+    stopSpinner();
   } catch {
+    stopSpinner("failed");
     return `Failed to download ${name}@${version} from ${url}`;
```
Comment (folex) [inline src/lib/rust.ts:299]: Why no error in the message? `catch` should have caught some exception, right?

File: src/lib/execPromise.ts [hunk 11]
```diff
@@ -126,32 +140,39 @@ export const execPromise = ({
...
-        );
+        res({
+          expectedErrorMessage: `\n${color.yellow(
+            MARINE_CARGO_DEPENDENCY
+          )} requires ${color.yellow(
+            "build-essential"
+          )} to be installed. Please install it and try again.\nOn debian-based systems (e.g. Ubuntu) you can install it using this command:\n\n${color.yellow(
+            "sudo apt install build-essential"
+          )}\n`,
+        });
```
Comment (folex) [inline src/lib/execPromise.ts:151]: is that a structure? is that a multiline string? Very hard to follow all the `)} )} )}` on new lines.

Maybe create a string as a separate variable, and then do
```
let message = `\n${color.yellow(...`
res({expectedErrorMessage: message})
```

would be much easier to read
```

File: src/lib/execPromise.ts [hunk 12]
```diff
@@ -111,7 +120,12 @@ export const execPromise = ({
         stopSpinner(color.red("failed"));
       }
 
-      rej(new Error(`${printOutput ? "" : stderr}${error}\n${debugInfo}`));
+      res({
+        errorMessage: `${getCommandFailedMessage()}${getErrorMessage(
+          printOutput,
+          stderr
+        )}${error}`,
```
Comment (folex) [inline src/lib/execPromise.ts:127]: same here. Suggest to move message to a separate variable to make it easier to read

File: src/lib/execPromise.ts (hunk 3 above)
Comment (folex) [reply src/lib/execPromise.ts:88]: ditto: suggest to move message to a separate variable to make it easier to read

File: src/lib/execPromise.ts [hunk 13]
```diff
@@ -37,33 +41,40 @@ export type ExecPromiseArg = {
...
     startSpinner(spinnerMessage);
   }
 
-  const debugInfo = `Debug info:\n${commandToDisplay}\n`;
-
-  return new Promise<string>((res, rej): void => {
+  const result = await new Promise<
+    | { errorMessage: string }
+    | { expectedErrorMessage: string }
+    | { success: string }
```
Comment (folex) [inline src/lib/execPromise.ts:76]:
```suggestion
    | { stdout: string }
```

File: src/lib/execPromise.ts [hunk 14]
```diff
@@ -37,33 +41,40 @@ export type ExecPromiseArg = {
...
-export const execPromise = ({
+export const execPromise = async ({
   command,
   args,
   flags,
   options,
   spinnerMessage,
   timeout,
-  printOutput = false,
+  printOutput: printOutputArg = false,
```
Comment (folex) [inline src/lib/execPromise.ts:51]: why `printOutputArg` starts from a lower case? looks strange

and why do you need explicit type here anyway?

File: src/lib/execPromise.ts (hunk 1 above)
Comment (shamsartem) [reply src/lib/execPromise.ts:60]: if you use ./bin/dev then it is development. Otherwise it's not
It's just so while developing you can see the actual args and in prod secret keys are not leaked
I think now that we don't use aqua cli this part can even be removed
I will do it

File: src/lib/execPromise.ts (hunk 2 above)
Comment (shamsartem) [reply src/lib/execPromise.ts:66]: typescript api says it can be null. I just handle this possibility. Don't know how to actually achieve this

File: src/lib/execPromise.ts [hunk 15]
```diff
@@ -17,16 +17,20 @@
...
+import { isInteractive } from "./commandObj.js";
 import { IS_DEVELOPMENT, MARINE_CARGO_DEPENDENCY } from "./const.js";
 import { type Flags, flagsToArgs } from "./helpers/flagsToArgs.js";
 import { startSpinner, stopSpinner } from "./helpers/spinner.js";
 
 type SpawnParams = Parameters<typeof spawn>;
 
+const getErrorMessage = (printOutput: boolean, stderr: string) => {
+  return printOutput ? "" : `. Reason:\n${stderr}`;
+};
```
[the remark this answers is not in the document]
Comment (shamsartem) [reply src/lib/execPromise.ts:32]: This shouldn't be resued anywhere in my opinion. This is to specific for this part of the codebase

File: src/lib/execPromise.ts (hunk 14 above)
Comment (shamsartem) [reply src/lib/execPromise.ts:51]: it's not a type it's a variable

File: src/lib/execPromise.ts (hunk 3 above)
Comment (shamsartem) [reply src/lib/execPromise.ts:88]: The command will be inside getCommandFailedMessage. I can move this to a variable. Not sure if it really impores readability at all

File: src/lib/marineCli.ts [hunk 16]
```diff
@@ -62,16 +63,28 @@ export const initMarineCli = async (
...
+            : getMessageWithKeyValuePairs(message, keyValuePairs),
+        options: { cwd },
+        printOutput,
+      });
+    } catch (e) {
+      if (
+        isObject(e) &&
+        hasKey("message", e) &&
+        typeof e["message"] === "string"
+      ) {
```
Comment (folex) [inline src/lib/marineCli.ts:83]: Maybe have a custom type for errors and do `instance of MyError`?

File: src/lib/execPromise.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (folex) [reply src/lib/execPromise.ts:32]: ok then

File: src/lib/execPromise.ts (hunk 4 above)
Comment (shamsartem) [reply src/lib/execPromise.ts:175]: expected error message (CLIError) is the same thing that happens when you do commandObj.error - it's just an error that is expected to be encountered by the user of the CLI. When expected error happens CLI doesn't print stack trace and also it is sent to Countly in a different way

File: src/lib/execPromise.ts (hunk 14 above)
Comment (folex) [reply src/lib/execPromise.ts:51]: I don't understand. What does `printOutput: printOutputArg = false` mean?

Comment (shamsartem) [reply src/lib/execPromise.ts:51]: printOutput - is destructuring; printOutput: printOutputArg - is renaming of the destructured value; printOutput: printOutputArg = false - is a default value in case the destructured value is undefined

File: src/lib/execPromise.ts (hunk 4 above)
Comment (folex) [reply src/lib/execPromise.ts:175]: so that's basically a non-fatal error? I've seen languages separate Fatal error from others.

maybe name fields as `fatalError` and `nonFatalError`? not sure

File: src/lib/rust.ts (hunk 10 above)
Comment (shamsartem) [reply src/lib/rust.ts:299]: good catch. Will add

File: src/lib/execPromise.ts (hunk 4 above)
Comment (shamsartem) [reply src/lib/execPromise.ts:176]: they are all fatal, meaining that after they happen cli stops execution. The difference really is if the errors are expected or not. If they are expected then the error message is nice and user knows what to do next and we don't need to fix anything. But unexpected errors are something that should not normally happen

File: src/lib/execPromise.ts (hunk 14 above)
Comment (folex) [reply src/lib/execPromise.ts:51]: omg :) ok

File: src/lib/execPromise.ts (hunk 4 above)
Comment (folex) [reply src/lib/execPromise.ts:176]: okay. Let's at least document it in the code?

File: src/lib/marineCli.ts (hunk 16 above)
Comment (shamsartem) [reply src/lib/marineCli.ts:83]: It doesn't matter what is the type of the error here I think. if error has a message - then show it. The change is about treating marine errors as exepected errors so we don't show stack traces when marine command fails

File: src/lib/marineCli.ts (hunk 8 above)
Comment (shamsartem) [reply src/lib/marineCli.ts:84]: here is different. Error objects by convention have `message`. errorMessage and expectedErrorMessage is just a way to resolve promise and distinguish between the way it was resolved. I think we can avoid errorMessage, expectedErrorMessage and use instance of in execPromise instead, that would be better

File: src/lib/marineCli.ts (hunk 16 above)
Comment (shamsartem) [reply src/lib/marineCli.ts:83]: I imporoved that part, but this "message" will never change. This is what happens when you do new Error() in js - it always creates an object with a "message" property in it. CLIError extends default js error but is the same

Comment (folex) [reply src/lib/marineCli.ts:83]: if you do `instanceof CLIError`, then you can access message as a field, via `.`, right?

File: src/lib/execPromise.ts (hunk 4 above)
Comment (shamsartem) [reply src/lib/execPromise.ts]: I imporved the code - check it out and tell me if any additional documentation is required

File: src/lib/marineCli.ts (hunk 16 above)
Comment (shamsartem) [reply src/lib/marineCli.ts]: It's true, but it doesn't matter what kind of error we have. Any error with a message is good. I added a genral check for that without using isntanceof, please check it out

File: src/lib/execPromise.ts [hunk 17]
```diff
@@ -126,32 +136,31 @@ export const execPromise = ({
...
+        const commandFailedMessage = getCommandFailedMessage(code);
+        const errorMessage = getErrorMessage(printOutput, stderr);
+        res(new Error(`${commandFailedMessage}${errorMessage}`));
       }
 
       res(stdout);
     });
   });
+
+  if (result instanceof CLIError || result instanceof Error) {
```
Comment (folex) [inline src/lib/execPromise.ts:156]: Aren't all `CLIError` are instances of `Error`?

File: src/lib/rust.ts [hunk 18]
```diff
@@ -285,19 +286,19 @@ const tryDownloadingBinary = async ({
...
-  try {
-    await downloadFile(binaryPath, url);
-  } catch {
-    return `Failed to download ${name}@${version} from ${url}`;
-  }
+  startSpinner(
+    `Downloading ${name}@${version} binary to ${replaceHomeDir(
+      dependencyDirPath
+    )}`
+  );
```
Comment (folex) [inline src/lib/rust.ts:293]:
```suggestion
  startSpinner(`Downloading ${name}@${version} binary to ${replaceHomeDir(dependencyDirPath)}`);
```

to me it's much easier to read, but if for you it's not, I'm ok with either

File: src/lib/rust.ts [hunk 19]
```diff
@@ -285,19 +286,19 @@ const tryDownloadingBinary = async ({
...
 
   try {
-    await execPromise({
-      command: "chmod",
-      args: ["+x", binaryPath],
-    });
-  } catch {
-    return `Failed to make ${name}@${version} executable by running chmod +x '${binaryPath}'`;
+    await downloadFile(binaryPath, url);
+    stopSpinner();
```
Comment (folex) [inline src/lib/rust.ts:297]: Maybe move `stopSpinner` to `finally`?

File: src/lib/typeHelpers.ts [hunk 20]
```diff
@@ -106,3 +106,13 @@ export function getIsStringUnion<T extends string>(
     });
   };
 }
+
+export function isObjectWithStringMessage(
+  unknown: unknown
+): unknown is { message: string } {
```
Comment (folex) [inline src/lib/typeHelpers.ts:112]: Why not `unknown is Error`?

**Review by folex — APPROVED** [review]

File: src/lib/execPromise.ts (hunk 17 above)
Comment (shamsartem) [reply src/lib/execPromise.ts:157]: Yeah. It is. Wanted to be overprotective in case CLIError implementation changes for some reason but seems like typescript would catch that issue for me. So it's good to remove additional check

File: src/lib/typeHelpers.ts (hunk 20 above)
Comment (shamsartem) [reply src/lib/typeHelpers.ts:112]: Because it's not true. I am not checking if it's an error here. That would be done with instanceof. I am thinking maybe you are right. I should just use instanceof. I am just inventing unnecessary generaliztions here

File: src/lib/rust.ts (hunk 18 above)
Comment (shamsartem) [reply src/lib/rust.ts:293]: the formatiing is done with prettier, no easy ways to control it and it's a good thing - less to argue about on code reviews

Comment (folex) [reply src/lib/rust.ts:293]: yep, makes sense

File: src/lib/rust.ts (hunk 19 above)
Comment (shamsartem) [reply src/lib/rust.ts:297]: finally is for the cases when no matter error happened or not - you wanna do _the same thing_
here we do different things dependening on weateher error happened or not
if error happens - spinner is stopped with the message "fail"
if error does not happen - spinnner stops with the default message (in the spinner implementation that I previously used the message was "done")

so it looks like this

```
Downloading marine@0.1.1 binary to ~/.fluence/cargo/marine/0.1.1...failed
```

of


```
Downloading marine@0.1.1 binary to ~/.fluence/cargo/marine/0.1.1...done
```

but currently spinners are turned off

Comment (folex) [reply src/lib/rust.ts:297]: ah, missed that. ok!
